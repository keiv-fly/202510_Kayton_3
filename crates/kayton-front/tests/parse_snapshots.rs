use insta::assert_snapshot;
use kayton_front::tests_support::parse_str;

#[test]
fn parse_simple_function() {
    let source = r#"fn add(a, b):
    let total = a + b
    total
"#;
    let output = parse_str("add.ktn", source);
    assert_snapshot!(
        "simple_function",
        format!("{:#?}\n---\n{:#?}", output.module, output.diagnostics)
    );
}

#[test]
fn parse_if_else_suite() {
    let source = r#"fn classify(x):
    if x > 0:
        "positive"
    elif x < 0:
        "negative"
    else:
        "zero"
"#;
    let output = parse_str("classify.ktn", source);
    assert_snapshot!(
        "if_else_suite",
        format!("{:#?}\n---\n{:#?}", output.module, output.diagnostics)
    );
}

#[test]
fn parse_inline_if_expression() {
    let source = r#"fn absolute(x):
    let sign = if x < 0: -1 else: 1
    if sign == -1:
        -x
    else:
        x
"#;
    let output = parse_str("absolute.ktn", source);
    assert_snapshot!(
        "inline_if",
        format!("{:#?}\n---\n{:#?}", output.module, output.diagnostics)
    );
}

#[test]
fn parse_while_loop() {
    let source = r#"fn countdown(n):
    let current = n
    while current > 0:
        let _ = current
        let current = current - 1
        ()
    ()
"#;
    let output = parse_str("countdown.ktn", source);
    assert_snapshot!(
        "while_loop",
        format!("{:#?}\n---\n{:#?}", output.module, output.diagnostics)
    );
}
