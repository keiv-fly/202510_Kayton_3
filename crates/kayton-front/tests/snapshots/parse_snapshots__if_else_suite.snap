---
source: crates/kayton-front/tests/parse_snapshots.rs
assertion_line: 25
expression: "format!(\"{:#?}\\n---\\n{:#?}\", output.module, output.diagnostics)"
---
HirModule {
    id: HirId(
        1,
    ),
    items: [
        Function(
            HirFunction {
                id: HirId(
                    2,
                ),
                name: Symbol(
                    0,
                ),
                params: [
                    HirParam {
                        id: HirId(
                            3,
                        ),
                        name: Symbol(
                            1,
                        ),
                        span: Span {
                            source: SourceId(
                                1,
                            ),
                            start: 12,
                            end: 13,
                        },
                    },
                ],
                body: HirBlock {
                    id: HirId(
                        4,
                    ),
                    statements: [],
                    tail: Some(
                        If(
                            HirIf {
                                id: HirId(
                                    5,
                                ),
                                condition: Binary(
                                    HirBinary {
                                        id: HirId(
                                            6,
                                        ),
                                        op: Gt,
                                        lhs: Name(
                                            HirNameRef {
                                                id: HirId(
                                                    7,
                                                ),
                                                name: Symbol(
                                                    1,
                                                ),
                                                span: Span {
                                                    source: SourceId(
                                                        1,
                                                    ),
                                                    start: 23,
                                                    end: 24,
                                                },
                                            },
                                        ),
                                        rhs: Literal(
                                            Int(
                                                HirIntLiteral {
                                                    id: HirId(
                                                        8,
                                                    ),
                                                    value: "0",
                                                    span: Span {
                                                        source: SourceId(
                                                            1,
                                                        ),
                                                        start: 27,
                                                        end: 28,
                                                    },
                                                },
                                            ),
                                        ),
                                        span: Span {
                                            source: SourceId(
                                                1,
                                            ),
                                            start: 23,
                                            end: 28,
                                        },
                                    },
                                ),
                                then_branch: HirBlock {
                                    id: HirId(
                                        9,
                                    ),
                                    statements: [],
                                    tail: Some(
                                        Literal(
                                            String(
                                                HirStringLiteral {
                                                    id: HirId(
                                                        10,
                                                    ),
                                                    value: "positive",
                                                    span: Span {
                                                        source: SourceId(
                                                            1,
                                                        ),
                                                        start: 38,
                                                        end: 48,
                                                    },
                                                },
                                            ),
                                        ),
                                    ),
                                    span: Span {
                                        source: SourceId(
                                            1,
                                        ),
                                        start: 28,
                                        end: 53,
                                    },
                                },
                                else_branch: Some(
                                    HirBlock {
                                        id: HirId(
                                            11,
                                        ),
                                        statements: [],
                                        tail: Some(
                                            Binary(
                                                HirBinary {
                                                    id: HirId(
                                                        12,
                                                    ),
                                                    op: Lt,
                                                    lhs: Name(
                                                        HirNameRef {
                                                            id: HirId(
                                                                13,
                                                            ),
                                                            name: Symbol(
                                                                1,
                                                            ),
                                                            span: Span {
                                                                source: SourceId(
                                                                    1,
                                                                ),
                                                                start: 58,
                                                                end: 59,
                                                            },
                                                        },
                                                    ),
                                                    rhs: Literal(
                                                        Int(
                                                            HirIntLiteral {
                                                                id: HirId(
                                                                    14,
                                                                ),
                                                                value: "0",
                                                                span: Span {
                                                                    source: SourceId(
                                                                        1,
                                                                    ),
                                                                    start: 62,
                                                                    end: 63,
                                                                },
                                                            },
                                                        ),
                                                    ),
                                                    span: Span {
                                                        source: SourceId(
                                                            1,
                                                        ),
                                                        start: 58,
                                                        end: 63,
                                                    },
                                                },
                                            ),
                                        ),
                                        span: Span {
                                            source: SourceId(
                                                1,
                                            ),
                                            start: 58,
                                            end: 63,
                                        },
                                    },
                                ),
                                span: Span {
                                    source: SourceId(
                                        1,
                                    ),
                                    start: 20,
                                    end: 53,
                                },
                            },
                        ),
                    ),
                    span: Span {
                        source: SourceId(
                            1,
                        ),
                        start: 14,
                        end: 88,
                    },
                },
                span: Span {
                    source: SourceId(
                        1,
                    ),
                    start: 0,
                    end: 88,
                },
            },
        ),
    ],
    interner: SymbolInterner {
        strings: {
            "classify",
            "x",
        },
    },
}
---
[
    Diagnostic {
        message: "expected expression",
        span: Span {
            source: SourceId(
                1,
            ),
            start: 63,
            end: 64,
        },
        severity: Error,
        notes: [],
    },
    Diagnostic {
        message: "expected expression",
        span: Span {
            source: SourceId(
                1,
            ),
            start: 73,
            end: 73,
        },
        severity: Error,
        notes: [],
    },
    Diagnostic {
        message: "expected `fn` or `let`",
        span: Span {
            source: SourceId(
                1,
            ),
            start: 88,
            end: 92,
        },
        severity: Error,
        notes: [],
    },
    Diagnostic {
        message: "expected `fn` or `let`",
        span: Span {
            source: SourceId(
                1,
            ),
            start: 102,
            end: 102,
        },
        severity: Error,
        notes: [],
    },
]
